modules/standard/DynamicIters.chpl

```diff
@@ -154,6 +154,113 @@ where tag == iterKind.follower
   }
 }
 
+//************************* Multidimensional Dynamic iterator
+/*
+
+  :arg c: The domain to iterate over. The rank of the domain must be greater 
+          than one. If using a 1d domain, the user should use the existing
+          dynamic iterators (use DynamicIters). 
+  :type c: `domain`
+
+  :arg chunkSize: The size of chunks to be yielded to each thread. Must be
+                  greater than zero
+  :type chunkSize: `int`
+
+  :arg numTasks: The number of tasks to use. Must be >= zero. If this argument
+                 has the value 0, it will use the value indicated by
+                 ``dataParTasksPerLocale``.
+  :type numTasks: `int`
+
+  :arg parDim: The index of the dimension to parallelize across. Must be > 0.
+                Must be <= the rank of the domain ``c``. Defaults to 1. 
+  :type parDim: `int`
+
+  :yields: Slices of the domain ``c``
+
+  Given an input domain ``c``, each task is assigned slices of ``c``. The
+  chunks each have ``chunkSize`` slices in them (or the remaining iterations
+   if there are fewer than ``chunkSize``). This continues until there are no 
+   remaining iterations in the dimension of ``c`` indicated by ``parDim``. 
+
+  LIMITATIONS:
+  user can select parDim, but only if they also define numTasks (even to 0)
+    should parDim be a non-default param?
+  explicitly types the ranges (so they will likely only work with ints)
+  May or may not properly lead/follow with other domain iterators
```

Comment:
I can add a call to the range iterator to handle one dimensional domains.

I used slices instead of indices because if the function is passed the domain {1..10, 3..4, 5..6} with chunk size one, it will return (1..1, 3..4, 5..6), (2..2, 3..4, 5..6), etc., and  I wanted to be clear that it is yielding a tuple of ranges. I also was originally returning a subdomain instead of a tuple of indices because I was confused about how leader/follower iterators for domains worked.

I have a test for leading/following with the default domain iterator, but don’t know what further testing is needed to prove that it works correctly with all of the rectangular domain iterators all of the time.

From: Ben Harshbarger <notifications@github.com>
Reply-To: chapel-lang/chapel <reply@reply.github.com>
Date: Tuesday, June 7, 2016 at 9:40 AM
To: chapel-lang/chapel <chapel@noreply.github.com>
Cc: Laura Delaney <ldelaney@cray.com>, Author <author@noreply.github.com>
Subject: Re: [chapel-lang/chapel] Add a dynamic iterator for domains with rank 2 or higher (#3966)


In modules/standard/DynamicIters.chpl<https://github.com/chapel-lang/chapel/pull/3966#discussion_r66107061>:

> +  :arg parDim: The index of the dimension to parallelize across. Must be > 0.

> +                Must be <= the rank of the domain ``c``. Defaults to 1.

> +  :type parDim: `int`

> +

> +  :yields: Slices of the domain ``c``

> +

> +  Given an input domain ``c``, each task is assigned slices of ``c``. The

> +  chunks each have ``chunkSize`` slices in them (or the remaining iterations

> +   if there are fewer than ``chunkSize``). This continues until there are no

> +   remaining iterations in the dimension of ``c`` indicated by ``parDim``.

> +

> +  LIMITATIONS:

> +  user can select parDim, but only if they also define numTasks (even to 0)

> +    should parDim be a non-default param?

> +  explicitly types the ranges (so they will likely only work with ints)

> +  May or may not properly lead/follow with other domain iterators

Two things:

You can name the argument 'parDim' to give it a value, like this:

var dom = {1..5, 1..10};

forall i in dynamic(dom, 1, parDIm=2) do ...

Second, I think by the end of this you should be able to say that this iterator should lead and follow with rectangular domains correctly, and with any domains that interact nicely with rectangular domains.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/chapel-lang/chapel/pull/3966/files/9ab730317465d794ab06e36f92577d2413902cbc#r66107061>, or mute the thread<https://github.com/notifications/unsubscribe/APJKMQxrGEOh1o8Mc9KFwozke305EznWks5qJZ8JgaJpZM4IvbXW>.